app/ante/eip712.go

```diff
@@ -0,0 +1,292 @@
+			tx.GetTimeoutHeight(),
+			legacytx.StdFee{
+				Amount: tx.GetFee(),
+				Gas:    tx.GetGas(),
+			},
+			msgs, tx.GetMemo(),
+		)
+
+		var (
+			chainID uint64
+			err     error
+
+			typedData apitypes.TypedData
+			sigHash   []byte
+
+			feePayer    sdk.AccAddress
+			feePayerSig []byte
+		)
+
+		signerChainID, err := ethermint.ParseChainID(signerData.ChainID)
+		if err != nil {
+			return sdkerrors.Wrapf(err, "failed to parse chainID: %s", signerData.ChainID)
+		}
+
+		txWithExtensions, ok := tx.(authante.HasExtensionOptionsTx)
+		if !ok {
+			return fmt.Errorf("tx doesnt contain any extensions")
+		}
+		opts := txWithExtensions.GetExtensionOptions()
+		if len(opts) != 1 {
+			return fmt.Errorf("tx doesnt contain expected amount of extension options")
+		}
+
+		var optIface ethermint.ExtensionOptionsWeb3TxI
+
+		if err := ethermintCodec.UnpackAny(opts[0], &optIface); err != nil {
+			return sdkerrors.Wrap(err, "failed to proto-unpack ExtensionOptionsWeb3Tx")
+		}
+
+		extOpt, ok := optIface.(*ethermint.ExtensionOptionsWeb3Tx)
+		if !ok {
+			return sdkerrors.Wrap(sdkerrors.ErrInvalidChainID, "unknown extension option")
+		}
+
+		if extOpt.TypedDataChainID != signerChainID.Uint64() {
+			return sdkerrors.Wrap(sdkerrors.ErrInvalidChainID, "invalid chainID")
+		}
+
+		chainID = extOpt.TypedDataChainID
+
+		if len(extOpt.FeePayer) == 0 {
+			return fmt.Errorf("no feePayer on ExtensionOptionsWeb3Tx")
+		}
+		feePayer, err = sdk.AccAddressFromBech32(extOpt.FeePayer)
+		if err != nil {
+			return sdkerrors.Wrap(err, "failed to parse feePayer from ExtensionOptionsWeb3Tx")
+		}
+
+		feePayerSig = extOpt.FeePayerSig
+		if len(feePayerSig) == 0 {
+			return fmt.Errorf("no feePayerSig provided in ExtensionOptionsWeb3Tx")
+		}
+
+		feeDelegation := &eip712.FeeDelegationOptions{
+			FeePayer: feePayer,
+		}
+
+		typedData, err = eip712.WrapTxToTypedData(ethermintCodec, chainID, msgs[0], txBytes, feeDelegation)
+		if err != nil {
+			return sdkerrors.Wrap(err, "failed to pack tx data in EIP712 object")
+		}
+
+		sigHash, err = eip712.ComputeTypedDataHash(typedData)
+		if err != nil {
+			return err
+		}
+
+		if len(feePayerSig) != ethcrypto.SignatureLength {
+			return fmt.Errorf("signature length doesn't match typical [R||S||V] signature 65 bytes")
+		}
+
+		// Remove the recovery offset if needed (ie. Metamask eip712 signature)
+		if feePayerSig[ethcrypto.RecoveryIDOffset] == 27 || feePayerSig[ethcrypto.RecoveryIDOffset] == 28 {
+			feePayerSig[ethcrypto.RecoveryIDOffset] -= 27
+		}
+
+		feePayerPubkey, err := secp256k1.RecoverPubkey(sigHash, feePayerSig)
+		if err != nil {
+			return sdkerrors.Wrap(err, "failed to recover delegated fee payer from sig")
+		}
+
+		ecPubKey, err := ethcrypto.UnmarshalPubkey(feePayerPubkey)
+		if err != nil {
+			return sdkerrors.Wrap(err, "failed to unmarshal recovered fee payer pubkey")
+		}
+
+		pk := &ethsecp256k1.PubKey{
+			Key: ethcrypto.CompressPubkey(ecPubKey),
+		}
+
+		if !pubKey.Equals(pk) {
```

Comment:
This may not be needed, because the cosmos pubkey is used when validating the `feePayerSig` at the `sepc256k1.VerifySignature`.
But having this check here makes the error to be descriptive and we can stop the validation early in case of error.